### DISCUSSION TITLE
Refactoring of HyperbolicGenerator

### DISCUSSION THREAD

**Discussion by manpen** [networkit/networkit#695] [Ideas]
Currently we have two generation algorithms for Hyperbolic Random Graphs (QuadTree and Band-based). Both are implemented in one class which also selects the best algorithm. There are a few issues:
- There are a bunch of tuning parameters (`setLeafCapacity`, `setBalance`, `setTheoreticalSplit`) which only affect the QuadTree
- There are methods (`getElapsedMilliseconds`) which only work with the band-based implementation

None of this is documented.

Making things worse, I want to add a third generator (and there's a further very interesting generator). In any case, there will be partial deprecation of older algorithms. Hence, I want to move each generator into its own class.

Two questions:
- Do we want to expose these algorithms to the user?
Then we have classes like `HyperbolicGeneratorBand` and `HyperbolicGeneratorQuadTree` which inherit from `StaticGenerator` and are also exported to Python. Alternatively, we treat the generators as implementation-specifics and do not export them. In any case the current `HyperbolicGenerator` would act as a dispatcher in order not to break old user code.

The first option is nicer for developers (as we can easily benchmark specific algorithms). But to users this should be of no concern and we should be able to switch the generators, as long as the results remain equivalent. For this reason I tent to keep the implementations private.

- How do we deal with implementation specific tuning parameters. I suggest that we drop them (again: I user should not care, and it stops us from switching algos later on). Hence, I suggest to mark them deprecated and if necessary make them dummys (still callable, but no effect).

**Comment by manpen** [comment]
Btw: the generator pulls `std` members into NetworKit's namespace at several places, e.g.,

```cpp
//  networkit/include/networkit/generators/quadtree/QuadNode.hpp 
using std::vector;
using std::min;
using std::max;
using std::cos;

// networkit/include/networkit/generators/quadtree/QuadNodeCartesianEuclid.hpp 
using std::vector;
using std::min;
using std::max;
using std::cos;

...
```

I think it's a no-brainer that we should not do this and fixed this. Do you want to have a separate PR for this?

**Comment by avdgrinten** [comment]
Yes, I think we want to expose the algorithms to the user. In many cases, we have multiple algorithms for the same problem and a single algorithm does not always dominate all others. In fact, the situation that individual algorithms are exposed to users is already a reality in NetworKit. Thus, also in this case, we should not hide the algorithms.

I agree that there need to be dispatchers that hide the difficulty of choosing the best (or a good) algorithm for a given problem, for users that do not care about the algorithmic details. While accessing specific cases should be possible, we should also aim to make the common case easy.

**Comment by avdgrinten** [comment]
Regarding the `using std::<something>` declarations: yes, that looks like a clear anti-pattern, especially in header files.
